### PR TITLE
[Fix] Ballerina init command is not working properly on Mac OS

### PR DESCRIPTION
**Pull request by rnavagamuwa**
## Purpose
This PR is related to the issue [14079](https://github.com/ballerina-platform/ballerina-lang/issues/14079).

### CODE REVIEW

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 1]
```diff
@@ -67,8 +68,26 @@
     private boolean helpFlag;
 
     private static boolean isDirEmpty(final Path directory) throws IOException {
+
         try (DirectoryStream<Path> dirStream = Files.newDirectoryStream(directory)) {
-            return !dirStream.iterator().hasNext();
+            //Check whether the OS is Mac OS and the folder contains .DS_Store file
```
Comment (Shan1024) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]:
```suggestion
            // Check whether the OS is MacOS and the folder contains `.DS_Store` file.
```

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 2]
```diff
@@ -67,8 +68,26 @@
...
+            if (pathIterator.hasNext()) {
+
+                Path path = (Path) pathIterator.next();
+                Path filePath = path.getFileName();
+                if (filePath != null && filePath.toString().equals(ProjectDirConstants.MAC_OS_DS_STORE_FILE)) {
+
+                    return System.getProperty("os.name").toLowerCase(Locale.ENGLISH).contains("mac") &&
+                            !pathIterator.hasNext();
+                } else {
+
```
Comment (Shan1024) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: I think it is better to remove this newline.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 3]
```diff
@@ -67,8 +68,26 @@
...
+                    return false;
+                }
+
```
Comment (Shan1024) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: I think it is better to remove this newline.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 4]
```diff
@@ -161,7 +180,8 @@ public void execute() {
                         do {
                             out.print("Module for the service: (no module) ");
                             packageName = scanner.nextLine().trim();
-                        } while (!validatePkgName(projectPath, packageName));
+                        }
+                        while (!validatePkgName(projectPath, packageName));
```
Comment (Shan1024) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Better to revert this change. Can be applied to few `while` loops which has been modified below as well.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 5]
```diff
@@ -67,8 +68,26 @@
     private boolean helpFlag;
 
     private static boolean isDirEmpty(final Path directory) throws IOException {
+
```
Comment (Shan1024) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Shall we remove this empty newline which is added to all of the methods?

**Comment by hasithaa** [comment]
@NatashaWso2 @hemikak Can you check this PR?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 6]
```diff
@@ -67,8 +68,25 @@
...
         try (DirectoryStream<Path> dirStream = Files.newDirectoryStream(directory)) {
-            return !dirStream.iterator().hasNext();
+            //Check whether the OS is MacOS and the folder contains .DS_Store file
+            Iterator pathIterator = dirStream.iterator();
+            if (pathIterator.hasNext()) {
+
+                Path path = (Path) pathIterator.next();
+                Path filePath = path.getFileName();
+                if (filePath != null && filePath.toString().equals(ProjectDirConstants.MAC_OS_DS_STORE_FILE)) {
```
Comment (NatashaWso2) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: We can simplify this if condition as below:
```java
if (filePath != null && filePath.toString().equals(ProjectDirConstants.MAC_OS_DS_STORE_FILE)) {
   return true;
}
return false;
```

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 7]
```diff
@@ -67,8 +68,25 @@
...
+
+                    return System.getProperty("os.name").toLowerCase(Locale.ENGLISH).contains("mac") &&
+                            !pathIterator.hasNext();
```
Comment (NatashaWso2) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Why do we need to check this condition again? 
```java
 !pathIterator.hasNext()
```

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 8]
```diff
@@ -67,8 +68,25 @@
     private boolean helpFlag;
 
     private static boolean isDirEmpty(final Path directory) throws IOException {
+
         try (DirectoryStream<Path> dirStream = Files.newDirectoryStream(directory)) {
-            return !dirStream.iterator().hasNext();
+            //Check whether the OS is MacOS and the folder contains .DS_Store file
+            Iterator pathIterator = dirStream.iterator();
+            if (pathIterator.hasNext()) {
+
```
Comment (NatashaWso2) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Shall we remove this extra line?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ProjectDirConstants.java [hunk 9]
```diff
@@ -56,4 +56,7 @@ private ProjectDirConstants() {
     public static final String BALLERINA_SOURCE_ROOT = "ballerina.source.root";
 
     public static final String NIGHTLY_BUILD = "nightly.build";
+
+    public static final String MAC_OS_DS_STORE_FILE = ".DS_Store";
```
Comment (NatashaWso2) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ProjectDirConstants.java]: Shall we rename the variable as 'DS_STORE_FILE' without specifying the os name?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 10]
```diff
@@ -256,6 +277,7 @@ private Manifest createManifest(Scanner scanner, String createToml) {
      */
     @Override
     public String getName() {
+
```
Comment (NatashaWso2) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Shall we remove this empty newline?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 11]
```diff
@@ -264,6 +286,7 @@ public String getName() {
      */
     @Override
     public void printLongDesc(StringBuilder out) {
+
```
Comment (NatashaWso2) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Shall we remove this empty newline?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 12]
```diff
@@ -274,6 +297,7 @@ public void printLongDesc(StringBuilder out) {
      */
     @Override
     public void printUsage(StringBuilder out) {
+
```
Comment (NatashaWso2) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Shall we remove this empty newline?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 13]
```diff
@@ -67,8 +68,25 @@
     private boolean helpFlag;
 
     private static boolean isDirEmpty(final Path directory) throws IOException {
+
         try (DirectoryStream<Path> dirStream = Files.newDirectoryStream(directory)) {
-            return !dirStream.iterator().hasNext();
+            //Check whether the OS is MacOS and the folder contains .DS_Store file
+            Iterator pathIterator = dirStream.iterator();
+            if (pathIterator.hasNext()) {
```
Comment (NatashaWso2) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: We can fail fast here by returning 'true' if there are no files inside the directory?

```java
if (!pathIterator.hasNext()) { 
 return true;
}
```

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java (hunk 7 above)
Comment (rnavagamuwa) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Otherwise, we won't be able to detect if there are more files other than the `.DS_Store`. Assume there are two files and one is the `.DS_Store`. Iterator picks the `.DS_Store` as its first item. Then there should be a way to check the additional files. In this case `!pathIterator.hasNext()` will return false, but if this is not here then the whole function will return `true` which is not the case.

Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Okay got it :) So we can remove the check for the os and have only the check for the file name

Comment (rnavagamuwa) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Don't we have to check the OS as well? What happens if there is a manually created `.DS_Store` file in a Linux system?

Comment (NatashaWso2) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: There is a rare chance of that happening right :) Even if such a file was created we would want to generate the ballerina source file. So I think it would be better if we remove the check added for the os name.

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 14]
```diff
@@ -67,8 +68,21 @@
...
-            return !dirStream.iterator().hasNext();
+            //Check whether the OS is MacOS and the folder contains .DS_Store file
+            Iterator pathIterator = dirStream.iterator();
+            if (!pathIterator.hasNext()) {
+                return true;
+            }
+            Path path = (Path) pathIterator.next();
+            Path filePath = path.getFileName();
+            if (filePath != null && filePath.toString().equals(ProjectDirConstants.DS_STORE_FILE)) {
+
```
Comment (NatashaWso2) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Shall we remove all extra new lines added?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 15]
```diff
@@ -236,15 +251,17 @@ private Manifest createManifest(Scanner scanner, String createToml) {
             do {
                 out.print("Organization name: (" + defaultOrg + ") ");
                 orgName = scanner.nextLine().trim();
-            } while (!validateOrgName(orgName));
+            }
```
Comment (NatashaWso2) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Why have we moved the while to a new line?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 16]
```diff
@@ -236,15 +251,17 @@ private Manifest createManifest(Scanner scanner, String createToml) {
...
             // Set org-name
             manifest.setName(orgName.isEmpty() ? defaultOrg : orgName);
             String version;
             do {
                 out.print("Version: (" + DEFAULT_VERSION + ") ");
                 version = scanner.nextLine().trim();
                 version = version.isEmpty() ? DEFAULT_VERSION : version;
-            } while (!validateVersion(out, version));
+            }
+            while (!validateVersion(out, version));
```
Comment (NatashaWso2) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Why have we moved the while to a new line?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java [hunk 17]
```diff
@@ -67,8 +68,21 @@
...
+
         try (DirectoryStream<Path> dirStream = Files.newDirectoryStream(directory)) {
-            return !dirStream.iterator().hasNext();
+            //Check whether the OS is MacOS and the folder contains .DS_Store file
+            Iterator pathIterator = dirStream.iterator();
+            if (!pathIterator.hasNext()) {
+                return true;
+            }
+            Path path = (Path) pathIterator.next();
+            Path filePath = path.getFileName();
```
Comment (NatashaWso2) [inline cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: Shall we change this variable name to 'fileName' since we are referring to the file name?

File: cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java (hunk 7 above)
Comment (rnavagamuwa) [reply cli/ballerina-packerina/src/main/java/org/ballerinalang/packerina/cmd/InitCommand.java]: If that's the case will remove the OS check. Then the logic can be simplified to 

`return fileName != null && fileName.toString().equals(ProjectDirConstants.DS_STORE_FILE) && !pathIterator.hasNext();`

**Comment by NatashaWso2** [comment]
@rnavagamuwa can you resolve the conflicts?

**Comment by rnavagamuwa** [comment]
@NatashaWso2 All the conflicts have been resolved right?
